docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+# Quality of Service in Kubelet
+
+**Author**: Ananya Kumar (@AnanyaKumar)
+
+**Status**: Draft proposal; prototype in progress.
+
+This document presents the design of quality of service tiers for jobs scheduled on Kubelet, describes use cases, and lays out milestones for its development.
+
+## Motivation
+
+The Kubelet allocates resources to pods in a simple way. Right now there aren’t robust mechanisms in place to ensure that pods run reliably on an overloaded system, and nodes are not very efficiently utilized.
+
+In Kubernetes, some pods have limits specified, and some pods do not. **Pods with limits specified are not very well protected from pods without limits specified**. If a pod without an explicit memory limit goes overboard and uses lots of memory, other pods (with resource specifications) might be killed. This is bad, because users often want a way to launch containers that have resources guarantees, and that stay up reliably.
+
+On the other hand in the current system, **if all pods have specified limits, cluster utilization can be poor**. In the current implementation, Kubelet does not support over-committing pods. Very often, pods don’t use all the resources that they request, which leads to a lot of wasted resources. For example, we might have 4 pods, each reserving 1GB of memory in a system with 4GB memory, but only using 500MB of memory. Theoretically, we could slot in another 4 pods of this type, but Kubernetes will reject new pods.
+
+In this design document, we propose **implementing two tiers of pods** - one in which we guarantee resources (top-tier pods) and the second wherein we provide resources if they’re available and at a slightly lower quality (best-effort pods). Adding a best-effort tier of pods will increase node’s utilization in a Kubernetes cluster (in Borg, supporting best-effort pods increased utilization by  about 20%). We also implement **a policy of automatically classifying pods that don’t provide limits as best-effort**. This provides much better resource guarantees to pods with limits specified.
+
+## QoS Functionality
+
+We will introduce two tiers of pods: top-tier pods and best-effort pods. We will not support multiple levels or priorities.

```

Comment:
This is described in terms of pods, but we don't currently specify resource requirements of pods, only of containers. Do you plan to change that? If we're going to use cgroup_parent, we should.

We'd need to add resources at the pod level and define behavior for cases for all combinations of pod and container limits specified and not specified.